test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -1435,15 +1451,97 @@ public void rollingRestart() throws Exception {
     /**
      * Restarts all nodes in a rolling restart fashion ie. only restarts on node a time.
      */
-    public void rollingRestart(RestartCallback function) throws Exception {
-        restartAllNodes(true, function);
+    public synchronized void rollingRestart(RestartCallback callback) throws Exception {
+        int numNodesRestarted = 0;
+        for (NodeAndClient nodeAndClient : nodes.values()) {
+            callback.doAfterNodes(numNodesRestarted++, nodeAndClient.nodeClient());
+            restartNode(nodeAndClient, callback);
+        }
+    }
+
+    private void restartNode(NodeAndClient nodeAndClient, RestartCallback callback) throws Exception {
+        logger.info("Restarting node [{}] ", nodeAndClient.name);
+        if (activeDisruptionScheme != null) {
+            activeDisruptionScheme.removeFromNode(nodeAndClient.name, this);
+        }
+        // special case to allow stopping one node in a two node cluster and keep it functional
+        final boolean updateMinMaster = nodeAndClient.isMasterNode() && getMasterNodesCount() == 2 && autoManageMinMasterNodes;
+        if (updateMinMaster) {
+            updateMinMasterNodes(-1);
+        }
+        nodeAndClient.restart(callback, true, getNewMinMasterNodes(0));
+        if (activeDisruptionScheme != null) {
+            activeDisruptionScheme.applyToNode(nodeAndClient.name, this);
+        }
+        if (callback.validateClusterForming() || updateMinMaster) {
+            // we have to validate cluster size if updateMinMaster == true, because we need the
+            // second node to join in order to increment min_master_nodes back to 2.
+            validateClusterFormed();
+        }
+        if (updateMinMaster) {
+            updateMinMasterNodes(0);
```

Comment:
I'm not really sure I like passing in a delta to `updateMinMasterNodes`. I understand why you do it but maybe the number of master nodes is a better thing to pass even if it is more verbose.

I think it'd be more clear if this method looked like:
```java
int currentMasterNodes = getMasterNodeCount();
if (updateMinMaster) {
  // The master is going down but we want the cluster to function as though enough master nodes are present
  updateMinMasterNodes(currentMasterNodes - 1);
}
restart it
wait for it to be up if needed
if (updateMinMaster) {
  // The master is back so we can tell nodes about it.
  updateMinMasterNodes(currentMasterNodes);
}
```

I just get confused because when I work in deltas `-1` should balance with `+1` but when I work in absolutes I don't have to balance anything.